exe-common/src/network/api.rs

```diff
@@ -13,60 +10,19 @@ pub trait Api {
 
     /// Get one specific block (represented by its unique hash) from the
     /// network
-    fn get_block(&mut self, hash: HeaderHash) -> Result<Block>;
-
-    /// Fetch a finished epoch
-    ///
-    /// Note that calling this api too close to the windows of block instability will either likely
-    /// result in failure to get the pack, or unexpected result.
-    ///
-    /// In the case of hermes, an epoch is pack only after the latest block of the epoch
-    /// is considered immutable in the chain by being old enough; latest known block's date
-    /// is tip.date() - k where k is typically around 2200 blocks. TODO this is a configurable
-    /// parameter, so expect to be able to find the constant somewhere in the net::Config in the future.
-    fn fetch_epoch(&mut self, config: &net::Config, storage: &mut Storage, fep: FetchEpochParams) -> Result<FetchEpochResult>;
-}
+    fn get_block(&mut self, hash: &HeaderHash) -> Result<RawBlock>;
 
-/// Parameter for fetching a full epoch
-///
-/// the epoch_id refer to which full epoch we want to fetch,
-/// which is either used to make sure we're getting the
-/// right blocks in the native API, or directly as a way
-/// to get the right epoch in the http API.
-///
-/// The upper bound hash is used for the native API to have
-/// a way to query the blocks in a range. we always know
-/// the genesis hash (coming from the configuration) and
-/// the get_tip API (coming from querying a network node.
-///
-/// The previous_header_hash is the hash we used to make
-/// sure that the chain is actually valid; this hash
-/// which represent the latest known block in the previous epoch,
-/// should be equal to the genesis block of the new epoch
-/// that is going to be fetch
-///
-/// start_header_hash is only used in the native protocol
-/// when downloading the first block of the chain,
-/// since we can't start downloading at previous_header_hash
-/// which point to a non valid block
-#[derive(Debug)]
-pub struct FetchEpochParams {
-    pub epoch_id: EpochId,
-    pub start_header_hash: HeaderHash,
-    pub previous_header_hash: HeaderHash,
-    pub upper_bound_hash: HeaderHash
+    /// Get the blocks in the half-open interval (from, to] (if
+    /// inclusive = false) or [from, to] (if inclusive = true). FIXME:
+    /// the inclusive = true case is only needed because the native
+    /// protocol doesn't support fetching from the genesis_prev hash.
+    fn get_blocks(&mut self, from: &BlockRef, inclusive: bool, to: &BlockRef,
+                   got_block: &mut FnMut(&HeaderHash, &Block, &RawBlock) -> ());
```

Comment:
use a where clause to simplify this function declaration:

```rust
fn get_blocks<F>(&mut self, from: &BlockRef, inclusive: bool, to: &BlockRef, got_block: &mut F)
    where F: FnMut(&HeaderHash, &Block, &RawBlock) -> ()
```

this will help reading the code a bit better.

Reply:
why is there a mutable reference to this function ?
RawBlock = Block but serialised in cbor. What do you need the Block deserialised for?

Reply:
It passes a Block to the callee because the callee might need it (and indeed `net_sync()` does), and the caller already has it (since it performs some checks on the blocks and get the previous block in some cases), so it's more efficient to not have to deserialize it twice.